src/main/kotlin/org/rust/lang/core/psi/ext/RsQualifiedNamedElement.kt

```diff
@@ -91,55 +88,57 @@ data class RsQualifiedName private constructor(
-        val item = childItems.lastOrNull() ?: parentItem
-
-        // If it's link to crate, try to find the corresponding `lib.rs` file
-        if (item.type == CRATE) {
-            return findCrateRoot(item, psiManager, context)
+        /** `this` is absolute path even for relative links, conversion is done in [documentationAsHtml] */
+        val crateRoot = findCrateRoot(crateName, psiManager, context) ?: return null
+        return when (val itemType = itemType) {
+            CRATE -> crateRoot
+            FIELD, STRUCTFIELD -> resolveAsField(crateRoot)
+            else -> resolveAsPsiElement(itemType, crateRoot)
         }
+    }
 
-        // Otherwise, split search into two steps:
-        val project = context.project
-        // First: look for `RsQualifiedNamedElement` with the same name as expected,
-        // generate sequence of all possible reexports of this element and check
-        // if any variant has the same `RsQualifiedName`
-        var result = lookupInIndex(project, RsNamedElementIndex.KEY, { it }) { element ->
-            if (element !is RsQualifiedNamedElement) return@lookupInIndex null
-            val candidate = if (element is RsModDeclItem && item.type == MOD) {
-                element.reference.resolve() as? RsMod ?: return@lookupInIndex null
-            } else {
-                element
-            }
-            QualifiedNamedItem.ExplicitItem(candidate)
+    private fun resolveAsPsiElement(itemType: ItemType, crateRoot: RsFile): RsQualifiedNamedElement? {
+        val segments = buildList {
+            add("crate")
+            addAll(modSegments)
+            if (itemType != MOD) add(parentItem.name)
+            addAll(childItems.map { it.name })
         }
-
-        if (result == null) {
-            // Second: if the first step didn't find any suitable item,
-            // look over all direct item reexports (considering possible new alias),
-            // generate sequence of all possible reexports for each element and check
-            // if any variant has the same `RsQualifiedName`
-            result = lookupInIndex(project, RsReexportIndex.KEY, ReexportKey::ProducedNameKey) { useSpeck ->
-                if (useSpeck.isStarImport) return@lookupInIndex null
-                val candidate = useSpeck.path?.reference?.resolve() as? RsQualifiedNamedElement
-                    ?: return@lookupInIndex null
-                QualifiedNamedItem.ReexportedItem.from(useSpeck, candidate)
-            }
+        val pathText = segments.joinToString("::")
+        val ns = itemType.namespace() ?: return null
+        val path = RsCodeFragmentFactory(crateRoot.project).createPath(pathText, crateRoot, ns = ns) ?: return null
+        val results = path.reference?.multiResolve()?.filterIsInstance<RsQualifiedNamedElement>().orEmpty()
+        return when (results.size) {
+            0 -> null
+            1 -> results.single()
```

Comment:
Actually there was a bug in `itemType` comparison, I fixed it
Also removed the bang proc macro test since it doesn't supported now and instead will be fixed by #9377